neqo-transport/src/send_stream.rs

```diff
@@ -269,46 +267,53 @@ impl RangeTracker {
     /// Merge contiguous Acked ranges into the first entry (0). This range may
     /// be dropped from the send buffer.
     fn coalesce_acked_from_zero(&mut self) {
-        let acked_range_from_zero = self
+        while let Some((next_len, _)) = self
             .used
-            .get_mut(&0)
+            .get(&self.acked)
             .filter(|(_, state)| *state == RangeState::Acked)
-            .map(|(len, _)| *len);
-
-        if let Some(len_from_zero) = acked_range_from_zero {
-            let mut to_remove = SmallVec::<[_; 8]>::new();
-
-            let mut new_len_from_zero = len_from_zero;
-
-            // See if there's another Acked range entry contiguous to this one
-            while let Some((next_len, _)) = self
-                .used
-                .get(&new_len_from_zero)
-                .filter(|(_, state)| *state == RangeState::Acked)
-            {
-                to_remove.push(new_len_from_zero);
-                new_len_from_zero += *next_len;
-            }
-
-            if len_from_zero != new_len_from_zero {
-                self.used.get_mut(&0).expect("must be there").0 = new_len_from_zero;
-            }
-
-            for val in to_remove {
-                self.used.remove(&val);
-            }
+        {
+            let to_remove = self.acked;
+            self.acked += *next_len;
+            self.used.remove(&to_remove);
         }
     }
 
-    fn mark_range(&mut self, off: u64, len: usize, state: RangeState) {
+    pub fn mark_range(&mut self, off: u64, len: usize, state: RangeState) {
         if len == 0 {
             qinfo!("mark 0-length range at {}", off);
             return;
         }
 
         let subranges = self.chunk_range_on_edges(off, len as u64, state);
 
+        let mut first: bool = true;
         for (sub_off, sub_len, sub_state) in subranges {
+            // Check if the first range overlaps the acked-from-zero range and adjust
+            if first {
+                first = false;
+                if sub_off < self.acked {
```

Comment:
I think that you want

```suggestion
                if sub_off <= self.acked {
```

If `self.acked` is 0, then the first range that will increase the value will start from `0..`

I included that in the above.